dali/operators/decoder/nvjpeg/decoupled_api/nvjpeg_memory.cc

```diff
@@ -0,0 +1,123 @@
+// Copyright (c) 2020, NVIDIA CORPORATION. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "nvjpeg_memory.h"
+#include <nvjpeg.h>
+#include <cassert>
+#include <iostream> // TODO(janton) remove
+#include <map>
+#include <vector>
+#include <mutex>
+#include <shared_mutex>
+#include "dali/kernels/alloc.h"
+#include "dali/core/cuda_error.h"
+
+namespace dali {
+
+using kernels::memory::KernelUniquePtr;
+using kernels::AllocType;
+using kernels::memory::alloc_unique;
+using kernels::memory::Allocate;
+
+namespace nvjpeg_memory {
+
+struct Buffer {
+  Buffer(KernelUniquePtr<char> unq_ptr, AllocType type, size_t sz)
+      : ptr(std::move(unq_ptr)), alloc_type(type), size(sz) {}
+  KernelUniquePtr<char> ptr;
+  AllocType alloc_type;
+  size_t size;
+};
+
+using BufferPool = std::map<std::thread::id, std::map<AllocType, std::vector<Buffer>>>;
+std::shared_timed_mutex buffer_pool_mutex_;
+BufferPool buffer_pool_;
+
+void* GetBuffer(std::thread::id thread_id, AllocType alloc_type, size_t size) {
+  std::shared_lock<std::shared_timed_mutex> lock(buffer_pool_mutex_);
```

Comment:
It would be very useful if the user can get nvJPEG stats to set the right hint.

Reply:
Done

Reply:
done